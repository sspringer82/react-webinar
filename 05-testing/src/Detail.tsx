import React from 'react';
import { Book } from './Book';

type Props = {
  book: Book;
};

const Detail: React.FC<Props> = ({ book }) => {
  return (
    <div>
      <h1 data-testid="title">{book.title}</h1>
      <div data-testid="author">{book.author}</div>
      <div data-testid="isbn">{book.isbn}</div>
    </div>
  );
};

export default Detail;
