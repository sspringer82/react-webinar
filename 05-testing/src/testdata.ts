export const twoBooks = [
  {
    id: 1,
    isbn: '978-3-548-06409-4',
    title: 'The Lord of the Rings',
    author: 'J.R.R. Tolkien',
    price: 19.99,
    pages: 1216,
    year: 1954,
  },
  {
    id: 2,
    isbn: '978-3-86680-192-9',
    title: "The Hitchhiker's Guide to the Galaxy",
    author: 'Douglas Adams',
    price: 9.99,
    pages: 224,
    year: 1979,
  },
];

export const twoDuplicateBooks = [
  {
    id: 1,
    isbn: '978-3-548-06409-4',
    title: 'The Lord of the Rings',
    author: 'J.R.R. Tolkien',
    price: 19.99,
    pages: 1216,
    year: 1954,
  },
  {
    id: 1,
    isbn: '978-3-548-06409-4',
    title: 'The Lord of the Rings',
    author: 'J.R.R. Tolkien',
    price: 19.99,
    pages: 1216,
    year: 1954,
  },
];
