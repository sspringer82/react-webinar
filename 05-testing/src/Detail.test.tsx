import { describe, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { book } from './testdata';
import Detail from './Detail';

describe('Detail', () => {
  it('should render', () => {
    render(<Detail book={book} />);

    expect(screen.getByRole('heading')).toHaveTextContent(
      'The Lord of the Rings'
    );
    expect(screen.getByTestId('author')).toHaveTextContent('J.R.R. Tolkien');
    expect(screen.getByTestId('isbn')).toHaveTextContent('978-3-548-06409-4');
  });

  it.each([
    { id: 'title', value: 'The Lord of the Rings' },
    { id: 'author', value: 'J.R.R. Tolkien' },
    { id: 'isbn', value: '978-3-548-06409-4' },
  ])('ensure $id is rendered with $value', ({ id, value }) => {
    render(<Detail book={book} />);
    expect(screen.getByTestId(id)).toHaveTextContent(value);
  });

  it('should match a snapshot', () => {
    const { asFragment } = render(<Detail book={book} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
