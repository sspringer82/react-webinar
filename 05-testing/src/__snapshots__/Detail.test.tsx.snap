// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Detail > should match a snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      data-testid="title"
    >
      The Lord of the Rings
    </h1>
    <div
      data-testid="author"
    >
      J.R.R. Tolkien
    </div>
    <div
      data-testid="isbn"
    >
      978-3-548-06409-4
    </div>
  </div>
</DocumentFragment>
`;
